src/Arcus.Security.KeyVault/KeyVaultSecretProvider.cs

```diff
@@ -1,16 +1,69 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.Net;
 using System.Text;
 using System.Threading.Tasks;
-using Arcus.Security.Core;
+using Arcus.Security.Core.Exceptions;
+using Arcus.Security.Core.Interfaces;
+using Arcus.Security.KeyVault.Factories;
+using GuardNet;
+using Microsoft.Azure.KeyVault;
+using Microsoft.Azure.KeyVault.Core;
+using Microsoft.Azure.KeyVault.Models;
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
 
 namespace Arcus.Security.KeyVault
 {
+    /// <summary>
+    /// Secret key provider that connects to Azure Key Vault
+    /// </summary>
     public class KeyVaultSecretProvider : ISecretProvider
     {
-        public async Task<string> GetAsync(string name)
+        private readonly string _keyVaultUri;
+        private readonly KeyVaultClientFactory _keyVaultClientFactory;
+        private KeyVaultClient _keyVaultClient;
+
+        /// <summary>
+        /// Creates an Azure KeyVault Secret provider, connected to a specific Azure Key Vault
+        /// </summary>
+        /// <param name="keyVaultClientFactory">A <see cref="KeyVaultClientFactory"/> implementation that will be used to generate the KeyVaultClient</param>
+        /// <param name="keyVaultUri">The Uri of the Azure KeyVault you want to connect to.  <example>https://{vaultname}nebulus-iot.vault.azure.net/</example></param>
+        public KeyVaultSecretProvider(KeyVaultClientFactory keyVaultClientFactory, string keyVaultUri)
+        {
+            Guard.NotNullOrEmpty(keyVaultUri, nameof(keyVaultUri));
+            Guard.NotNull(keyVaultClientFactory, nameof(keyVaultClientFactory));
+            _keyVaultUri = keyVaultUri;
+            _keyVaultClientFactory = keyVaultClientFactory;
+        }
+
+        // Moving this implementation here, as KeyVaultClient cannot be mocked (https://github.com/Azure/azure-sdk-for-java/issues/1552)
+        // Therefore, constructor will not try to create the client 
```

Comment:
This should be moved to a method to perform the logic so that the client creation is done asynchronous, not synchronous

Reply:
fixed